lib/metadata/in_memory/backend.js

```diff
@@ -143,78 +149,54 @@ const metastore = {
                 return cb(errors.InvalidArgument);
             }
 
-            let numKeys = maxKeys;
             // If paramMaxKeys is undefined, the default parameter will set it.
             // However, if it is null, the default parameter will not set it.
+            let numKeys = maxKeys;
             if (numKeys === null) {
                 numKeys = defaultMaxKeys;
             }
 
             if (!metadata.keyMaps.has(bucketName)) {
                 return cb(errors.NoSuchBucket);
             }
-            const response = new ListBucketResult();
-            let keys = [];
+            const keys = [];
             metadata.keyMaps.get(bucketName).forEach((val, key) => {
-                keys.push(key);
+                if (marker === undefined || key > marker) {
+                    keys.push(key);
+                }
             });
             keys.sort();
             // If marker specified, edit the keys array so it
             // only contains keys that occur alphabetically after the marker
-            if (marker) {
-                keys = markerFilter(marker, keys);
-                response.Marker = marker;
-            }
-            // If prefix specified, edit the keys array so it only
-            // contains keys that contain the prefix
-            if (prefix) {
-                keys = prefixFilter(prefix, keys);
-                response.Prefix = prefix;
-            }
+            const filterParameters = {
+                delimiter,
+                start: prefix,
+                maxKeys: numKeys,
+                gt: marker,
+            };
+            const Ext = new algorithms.list.Delimiter(filterParameters, log);
+
             // Iterate through keys array and filter keys containing
             // delimiter into response.CommonPrefixes and filter remaining
             // keys into response.Contents
             for (let i = 0; i < keys.length; ++i) {
                 const currentKey = keys[i];
                 // Do not list object with delete markers
-                if (response.hasDeleteMarker(currentKey,
+                if (this._hasDeleteMarker(currentKey,
                     metadata.keyMaps.get(bucketName))) {
                     continue;
                 }
-                // If hit numKeys, stop adding keys to response
-                if (response.MaxKeys >= numKeys) {
-                    response.IsTruncated = true;
-                    response.NextMarker = keys[i - 1];
+                const objMD = metadata.keyMaps.get(bucketName).get(currentKey);
+                const value = JSON.stringify(objMD);
+                const obj = {
+                    key: currentKey,
+                    value,
+                };
```

Comment:
please add comment along the following lines:
// calling Ext.filter(obj) adds the obj to the Ext result if not filtered.
// Also, Ext.filter returns false when hit max keys.
// What a nifty function!